src/openfermion/ops/_code_operator.py

```diff
@@ -0,0 +1,298 @@
+    
+    As the occupation number of fermionic orbital is effectively binary,
+    a length-N vector (v) of binary number can be utilized to describe 
+    a configuration of a many-body fermionic state on N orbitals. 
+    An n-qubit product state configuration |w0> |w1> |w2> ... |wn-1>,
+    on the other hand is described by a length-n binary vector 
+    w=(w0, w1, ..., wn-1). To map a subset of N-Orbital Fermion states 
+    to n-qubit states we define a binary code, which consists of a 
+    (here: linear) encoding (e) and a (non-linear) decoding (d), such 
+    that for every v from that subset, w = e(v) is a length-n binary 
+    vector with d(w) = v.  This can be used to save qubits given a 
+    Hamiltonian that dictates such a subset, otherwise n=N. 
+    
+    Two binary codes (e,d) and (e',d') can construct a third code (e",d")
+    by two possible operations:
+    
+    Concatenation: (e",d") = (e,d) * (e',d')
+    which means e": v" -> e'( e(v") ) and d": w" -> d( d'(w") )
+    where n" = n' and N" = N, with n = N' as necessary condition. 
+    
+    Appendage: (e",d") = (e,d) + (e',d')
+    which means e": (v + v') -> e(v) + e'(v') and d": (w + w') -> d(w) + d'(w') 
+    where the addition is to be understood as appending two vectors together,
+    so N" = N' + N and n" = n + n'.
+    
+    Appending codes is particularly useful when considering segment codes or
+    segmented transforms. 
+    
+    A BinaryCode-instance is initialized by BinaryCode(A,d),
+    given the encoding (e) as n x N array or matrix-like nested lists A, 
+    such that e(v) = (A v) mod 2. The decoding d is an array or a list-like 
+    input of length N, which has entries either of type SymbolicBinary, or of 
+    valid type for an input of the SymbolicBinary-constructor. 
+    
+    The signs + and *, += and *= are overloaded to implement concatenation
+    and appendage on BinaryCode-objects.
+
+    NOTE: that multiplication of a BinaryCode with an integer yields a
+    multiple appending of the same code, the multiplication with another
+    BinaryCode their concatenation.
+    """
+
+    def __init__(self, encoding, decoding):
+        """ Initialization of a binary code.
+
+        Args: 
+            encoding (array or list): nested lists or binary 2D-array
+            decoding (array or list): list of SymbolicBinary, list-like or str
+
+        Raises: 
+            TypeError: non-list, array like encoding or decoding, unsuitable
+            symbolicBinary generators,
+            BinaryCodeError: in case of decoder/encoder size mismatch or decoder
+            size, qubits indexed mismatch
+        """
+
+        if not isinstance(encoding, (numpy.ndarray, list, tuple)):
+            raise TypeError('encoding must be a list, array or tuple .')
+
+        if not isinstance(decoding, (numpy.ndarray, list, tuple)):
+            raise TypeError('decoding must be a list, array or tuple .')
+
+        self.enc = scipy.sparse.csc_matrix(encoding)
```

Comment:
I really think these attributes should have a more complete name like "encoder/encoding" instead of "enc". There is something slightly arbitrary about such abbreviations which makes it a little less intuitive to anticipate how things are named without digging through the docs.